Test failure in CI build 9357

The following test appears to have failed:

[#9357](https://circleci.com/gh/cockroachdb/cockroach/9357):

```
I1113 03:29:27.861174 1014 gossip/gossip.go:172  gossiping node descriptor node_id:3 address:<network_field:"localhost" address_field:"3" > attrs:<> 
I1113 03:29:27.862243 1014 storage/queue.go:193  no system config available. skipping...
I1113 03:29:28.940092 1014 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
I1113 03:29:28.940248 1014 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
I1113 03:29:28.940318 1014 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- FAIL: TestRaftLogQueue (1.09s)
	testing.go:157: storage/client_raft_log_queue_test.go:93: condition failed to evaluate within 1s: storage/client_raft_log_queue_test.go:90: raft log has not been truncated yet, currentIndex:11 originalIndex:11
=== RUN   TestStoreRecoverFromEngine
I1113 03:29:28.940804 1014 gossip/gossip.go:172  gossiping node descriptor node_id:1 address:<network_field:"localhost" address_field:"1" > attrs:<> 
I1113 03:29:28.941509 1014 multiraft/multiraft.go:562  node 1 starting
I1113 03:29:28.941631 1014 raft/raft.go:409  [group 1] 1 became follower at term 5
I1113 03:29:28.941683 1014 raft/raft.go:218  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I1113 03:29:28.941755 1014 multiraft/multiraft.go:964  node 1 campaigning because initial confstate is [1]
I1113 03:29:28.941862 1014 storage/replica.go:1259  gossiping cluster id  from store 1, range 1
I1113 03:29:28.942010 1014 raft/raft.go:488  [group 1] 1 is starting a new election at term 5
I1113 03:29:28.942035 1014 raft/raft.go:422  [group 1] 1 became candidate at term 6
--
=== RUN   TestSplitSnapshotRace_SnapshotWins
--- SKIP: TestSplitSnapshotRace_SnapshotWins (0.00s)
	client_split_test.go:740: TODO(bdarnell): https://github.com/cockroachdb/cockroach/issues/3121
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.13s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	20.348s
=== RUN   TestBatchBasics
I1113 03:29:19.309201 991 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I1113 03:29:19.310244 991 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN   TestBatchMerge
I1113 03:29:19.312037 991 storage/engine/rocksdb.go:132  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/storage/client_raft_log_queue_test.go
+++ b/storage/client_raft_log_queue_test.go
@@ -20,11 +20,9 @@ package storage_test
 import (
 	"fmt"
 	"testing"
-	"time"
 
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/storage"
-	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/leaktest"
 )
 
@@ -73,29 +71,23 @@ func TestRaftLogQueue(t *testing.T) {
 
 	// Force a truncation check.
 	for _, store := range mtc.stores {
-		store.ForceRaftLogScan(t)
+		store.ForceRaftLogScanAndProcess(t)
 	}
 
-	// Wait until the firstIndex has increased indicating that the log
+	// Ensure that firstIndex has increased indicating that the log
 	// truncation has occurred.
-	var currentIndex uint64
-	util.SucceedsWithin(t, time.Second, func() error {
-		var err error
-		currentIndex, err = rep.FirstIndex()
-		if err != nil {
-			return err
-		}
-		if currentIndex <= originalIndex {
-			return util.Errorf("raft log has not been truncated yet, currentIndex:%d originalIndex:%d",
-				currentIndex, originalIndex)
-		}
-		return nil
-	})
+	currentIndex, err := rep.FirstIndex()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if currentIndex <= originalIndex {
+		t.Errorf("raft log has not been truncated yet, currentIndex:%d originalIndex:%d", currentIndex, originalIndex)
+	}
 
 	// Force a truncation check again to ensure that attempting to truncate an
 	// already truncated log has no effect.
 	for _, store := range mtc.stores {
-		store.ForceRaftLogScan(t)
+		store.ForceRaftLogScanAndProcess(t)
 	}
 
 	finalIndex, err := rep.FirstIndex()

--- a/storage/client_raft_log_queue_test.go
+++ b/storage/client_raft_log_queue_test.go
@@ -20,6 +20,7 @@ package storage_test
 import (
 	"fmt"
 	"testing"
+	"time"
 
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/storage"
@@ -28,12 +29,19 @@ import (
 
 // TestRaftLogQueue verifies that the raft log queue correctly truncates the
 // raft log.
-// TODO(bram): Add a new test case in which the raft leader differs from the
-// lease holder.
 func TestRaftLogQueue(t *testing.T) {
 	defer leaktest.AfterTest(t)
 
-	mtc := startMultiTestContext(t, 3)
+	mtc := new(multiTestContext)
+
+	// Turn off raft elections so the raft leader won't change out from under
+	// us in this test.
+	sc := storage.TestStoreContext
+	sc.RaftTickInterval = time.Hour * 24
+	sc.RaftElectionTimeoutTicks = 1000000
+	mtc.storeContext = &sc
+
+	mtc.Start(t, 3)
 	defer mtc.Stop()
 
 	// Write a single value to ensure we have a leader.
@@ -42,8 +50,13 @@ func TestRaftLogQueue(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	rep := mtc.stores[0].LookupReplica([]byte("a"), nil)
-	originalIndex, err := rep.FirstIndex()
+	// Get the raft leader (and ensure one exists).
+	desc := mtc.stores[0].LookupReplica([]byte("a"), nil).Desc()
+	raftLeaderRepl := mtc.getRaftLeader(desc.RangeID, time.Second)
+	if raftLeaderRepl == nil {
+		t.Fatalf("could not find raft leader replica for range %d", desc.RangeID)
+	}
+	originalIndex, err := raftLeaderRepl.FirstIndex()
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -56,19 +69,6 @@ func TestRaftLogQueue(t *testing.T) {
 		}
 	}
 
-	// Check the raft log and make sure it has advanced.
-	desc := rep.Desc()
-	raftStatus := mtc.stores[0].RaftStatus(desc.RangeID)
-	if raftStatus == nil {
-		t.Fatalf("the raft group doesn't exist for range %d", desc.RangeID)
-	}
-
-	for _, progress := range raftStatus.Progress {
-		if progress.Match <= originalIndex {
-			t.Fatalf("raft log has not advanced, match:%d originalIndex:%d", progress.Match, originalIndex)
-		}
-	}
-
 	// Force a truncation check.
 	for _, store := range mtc.stores {
 		store.ForceRaftLogScanAndProcess(t)
@@ -76,12 +76,13 @@ func TestRaftLogQueue(t *testing.T) {
 
 	// Ensure that firstIndex has increased indicating that the log
 	// truncation has occurred.
-	currentIndex, err := rep.FirstIndex()
+	afterTruncationIndex, err := raftLeaderRepl.FirstIndex()
 	if err != nil {
 		t.Fatal(err)
 	}
-	if currentIndex <= originalIndex {
-		t.Errorf("raft log has not been truncated yet, currentIndex:%d originalIndex:%d", currentIndex, originalIndex)
+	if afterTruncationIndex <= originalIndex {
+		t.Fatalf("raft log has not been truncated yet, afterTruncationIndex:%d originalIndex:%d",
+			afterTruncationIndex, originalIndex)
 	}
 
 	// Force a truncation check again to ensure that attempting to truncate an
@@ -90,12 +91,12 @@ func TestRaftLogQueue(t *testing.T) {
 		store.ForceRaftLogScanAndProcess(t)
 	}
 
-	finalIndex, err := rep.FirstIndex()
+	after2ndTruncationIndex, err := raftLeaderRepl.FirstIndex()
 	if err != nil {
 		t.Fatal(err)
 	}
-	if currentIndex != finalIndex {
-		t.Errorf("raft log was truncated again and it shouldn't have been, post 1st truncation index:%d post 2nd truncation index:%d",
-			currentIndex, finalIndex)
+	if afterTruncationIndex != after2ndTruncationIndex {
+		t.Fatalf("raft log was truncated again and it shouldn't have been, afterTruncationIndex:%d after2ndTruncationIndex:%d",
+			afterTruncationIndex, after2ndTruncationIndex)
 	}
 }

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -51,6 +51,7 @@ import (
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/stop"
+	"github.com/coreos/etcd/raft"
 	"github.com/gogo/protobuf/proto"
 )
 
@@ -567,6 +568,41 @@ func (m *multiTestContext) waitForValues(key roachpb.Key, d time.Duration, expec
 // replica sets.
 func (m *multiTestContext) expireLeaderLeases() {
 	m.manualClock.Increment(int64(storage.DefaultLeaderLeaseDuration) + 1)
+}
+
+// getRaftLeader returns the replica that is the current raft leader for the
+// specified rangeID.
+func (m *multiTestContext) getRaftLeader(rangeID roachpb.RangeID, d time.Duration) *storage.Replica {
+	var raftLeaderRepl *storage.Replica
+	util.SucceedsWithinDepth(1, m.t, d, func() error {
+		var latestTerm uint64
+		for _, store := range m.stores {
+			raftStatus := store.RaftStatus(rangeID)
+			if raftStatus == nil {
+				// Replica does not exist on this store or there is no raft
+				// status yet.
+				continue
+			}
+			if raftStatus.Term > latestTerm {
+				// If we find any newer term, it means any previous election is
+				// invalid.
+				raftLeaderRepl = nil
+				latestTerm = raftStatus.Term
+				if raftStatus.RaftState == raft.StateLeader {
+					var err error
+					raftLeaderRepl, err = store.GetReplica(rangeID)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+		if latestTerm == 0 || raftLeaderRepl == nil {
+			return util.Errorf("could not find a raft leader for range %s", rangeID)
+		}
+		return nil
+	})
+	return raftLeaderRepl
 }
 
 // getArgs returns a GetRequest and GetResponse pair addressed to

--- a/storage/queue.go
+++ b/storage/queue.go
@@ -319,8 +319,8 @@ func (bq *baseQueue) processLoop(clock *hlc.Clock, stopper *stop.Stopper) {
 	})
 }
 
+// processOne processes the next replica in the queue.
 func (bq *baseQueue) processOne(clock *hlc.Clock) {
-	start := time.Now()
 	bq.Lock()
 	repl := bq.pop()
 	bq.Unlock()
@@ -329,6 +329,13 @@ func (bq *baseQueue) processOne(clock *hlc.Clock) {
 		return
 	}
 
+	bq.processReplica(repl, clock)
+}
+
+// processReplica processes a single replica.
+// This should not be called externally to the queue.
+func (bq *baseQueue) processReplica(repl *Replica, clock *hlc.Clock) {
+	start := time.Now()
 	now := clock.Now()
 
 	// Load the system config.
@@ -347,7 +354,6 @@ func (bq *baseQueue) processOne(clock *hlc.Clock) {
 		}
 		return
 	}
-
 	if log.V(3) {
 		log.Infof("processing replica %s from %s queue...", repl, bq.name)
 	}
@@ -388,4 +394,15 @@ func (bq *baseQueue) pop() *Replica {
 func (bq *baseQueue) remove(index int) {
 	item := heap.Remove(&bq.priorityQ, index).(*replicaItem)
 	delete(bq.replicas, item.value.Desc().RangeID)
+}
+
+// DrainQueue locks the queue and processes the remaining queued replicas. It
+// processes the replicas in the order they're queued in, one at a time.
+// Exposed for testing only.
+func (bq *baseQueue) DrainQueue(clock *hlc.Clock) {
+	bq.Lock()
+	defer bq.Unlock()
+	for repl := bq.pop(); repl != nil; repl = bq.pop() {
+		bq.processReplica(repl, clock)
+	}
 }

--- a/storage/store.go
+++ b/storage/store.go
@@ -700,15 +700,18 @@ func (s *Store) DisableRaftLogQueue(disabled bool) {
 	s.raftLogQueue.SetDisabled(disabled)
 }
 
-// ForceRaftLogScan iterates over all ranges and enqueues any that need their
-// raft logs truncated. Exposed only for testing.
-func (s *Store) ForceRaftLogScan(t util.Tester) {
+// ForceRaftLogScanAndProcess iterates over all ranges and enqueues any that
+// need their raft logs truncated and then process each of them.
+// Exposed only for testing.
+func (s *Store) ForceRaftLogScanAndProcess(t util.Tester) {
+	// Add each range to the queue.
 	s.mu.Lock()
-	defer s.mu.Unlock()
-
 	for _, r := range s.replicas {
 		s.raftLogQueue.MaybeAdd(r, s.ctx.Clock.Now())
 	}
+	s.mu.Unlock()
+
+	s.raftLogQueue.DrainQueue(s.ctx.Clock)
 }
 
 // Bootstrap writes a new store ident to the underlying engine. To

